File: src/main/java/xyz/avarel/waffles/nodes/IfNode.java
package xyz.avarel.waffles.nodes;

import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.UnexpectedResultException;
import com.oracle.truffle.api.profiles.ConditionProfile;

public class IfNode extends WaffleNode {
    @Child
    private WaffleNode testNode;
    @Child
    private WaffleNode thenNode;
    @Child
    private WaffleNode elseNode;

    private final ConditionProfile conditionProfile = ConditionProfile.createBinaryProfile();

    public IfNode(WaffleNode testNode, WaffleNode thenNode, WaffleNode elseNode) {
        this.testNode = testNode;
        this.thenNode = thenNode;
        this.elseNode = elseNode;
    }

    @Override
    public Object execute(VirtualFrame virtualFrame) {
        if (this.conditionProfile.profile(testResult(virtualFrame))) {
            return thenNode.execute(virtualFrame);
        } else {
            return elseNode.execute(virtualFrame);
        }
    }

    private boolean testResult(VirtualFrame virtualFrame) {
        try {
            return testNode.executeBoolean(virtualFrame);
        } catch (UnexpectedResultException ignore) {
            return false;
        }
    }
}

File: src/main/kotlin/xyz/avarel/waffles/nodes/flow/IfNode.kt
package xyz.avarel.waffles.nodes.flow

import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.UnexpectedResultException
import com.oracle.truffle.api.profiles.ConditionProfile
import xyz.avarel.waffles.nodes.WaffleNode

class IfNode(
        @field:Child private var testNode: WaffleNode,
        @field:Child private var thenNode: WaffleNode,
        @field:Child private var elseNode: WaffleNode
) : WaffleNode() {
    private val conditionProfile = ConditionProfile.createBinaryProfile()

    override fun execute(virtualFrame: VirtualFrame): Any {
        return if (this.conditionProfile.profile(testResult(virtualFrame))) {
            thenNode.execute(virtualFrame)
        } else {
            elseNode.execute(virtualFrame)
        }
    }

    private fun testResult(virtualFrame: VirtualFrame): Boolean {
        return try {
            testNode.executeBoolean(virtualFrame)
        } catch (ignore: UnexpectedResultException) {
            false
        }

    }
}
